### PR TITLE
MAT-3931 upload MAT measure data to S3 bucket for measure transfer to…

### PR DESCRIPTION
**Pull request by adongare**
MAT-3931 upload MAT measure data to S3 bucket for measure transfer to MADiE

## Description

- Added AWS S3 SDK
- Implemented AWS S3 file upload for MAT measure transfer to MADiE
- Changelog to add a transferComplete flag to measure to mark measure transfer complete(default to false)
## Checklist

- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [ ] Tests are included and test edge cases.
- [x] Tests have been run locally and pass.

## Screenshots (if appropriate)

- [ ] None applicable

### CODE REVIEW

File: src/main/java/mat/dto/MeasureTransferDTO.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import java.io.Serializable;
+
+public class MeasureTransferDTO implements Serializable {
+    private String emailId;
+    private String harpId;
+    private ManageMeasureDetailModel manageMeasureDetailModel;
+    private String measureResourceJson;
+    private String libraryResourcesJson;
+
+    public String getEmailId() {
```
Comment (sb-cecilialiu) [inline src/main/java/mat/dto/MeasureTransferDTO.java:14]: Just curious, lombok is not available?

File: src/main/java/mat/server/util/MeasureTransferUtil.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        String objectKeyName = "measure_" + measureId;
+        String bucketName = getMeasureTransferS3BucketName();
+
+        ObjectMapper objectMapper = new ObjectMapper();
+        String transferJson = objectMapper.writeValueAsString(ob);
+        return buildAwsS3Client()
+                .putObject(bucketName, objectKeyName, transferJson);
+    }
+
+    public static final String getMeasureTransferS3Region() {
```
Comment (sb-cecilialiu) [inline src/main/java/mat/server/util/MeasureTransferUtil.java:33]: It's probably clearer, but it probably can be replaced by getSystemProperty() right?

File: src/main/java/mat/dto/MeasureTransferDTO.java (hunk 1 above)
Comment (adongare) [reply src/main/java/mat/dto/MeasureTransferDTO.java:14]: @sb-cecilialiu, that is a good point. I thought MAT doesn't have it but turns out it has. I'll replace it with Lombok annotations. Thank you!

File: src/main/java/mat/server/util/MeasureTransferUtil.java (hunk 2 above)
Comment (adongare) [reply src/main/java/mat/server/util/MeasureTransferUtil.java:33]: yeah maybe.... let me do it. I think doesn't add any value adding a separate method.

Comment (adongare) [reply src/main/java/mat/server/util/MeasureTransferUtil.java]: done

File: src/main/java/mat/server/util/MeasureTransferUtil.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        return AmazonS3ClientBuilder
+                .standard()
+                .withRegion(getMeasureTransferS3Region())
+                .withCredentials(new InstanceProfileCredentialsProvider(true))
+                .build();
+    }
+
+    public static final PutObjectResult uploadMeasureDataToS3Bucket(MeasureTransferDTO ob, String measureId)
+            throws JsonProcessingException {
+        String objectKeyName = "measure_" + measureId;
```
[the remark this answers is not in the document]
Comment (adongare) [reply src/main/java/mat/server/util/MeasureTransferUtil.java:25]: Not really. No one is going to extend override this class.

**Review by RohitKandimalla — DISMISSED** [review]

**Review by sb-cecilialiu — DISMISSED** [review]
Thank you so much Ashok! It's great work!

File: src/main/java/mat/dto/MeasureTransferDTO.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import java.io.Serializable;
+
+@Data
+@NoArgsConstructor
+public class MeasureTransferDTO implements Serializable {
+    private String emailId;
+    private String harpId;
+    private ManageMeasureDetailModel manageMeasureDetailModel;
+    private String measureResourceJson;
+    private String libraryResourcesJson;
```
Comment (jkotanchik-SB) [inline src/main/java/mat/dto/MeasureTransferDTO.java:16]:
```suggestion
    private String measureResourceElmJson;
    private String libraryResourcesElmJson;
```
The JSON should be the ELM.  Recommend renaming the member variable as above to add some clarity.

File: src/main/java/mat/server/util/MeasureTransferUtil.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkotanchik-SB) [reply src/main/java/mat/server/util/MeasureTransferUtil.java:25]: That is the point of adding `final`, to explicitly prevent such actions.

File: src/main/java/mat/dto/MeasureTransferDTO.java (hunk 4 above)
Comment (adongare) [reply src/main/java/mat/dto/MeasureTransferDTO.java:16]: oh wow, I can commit the change right here. I'll update it. Thanks Joe!

File: src/main/java/mat/server/util/MeasureTransferUtil.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    public static AmazonS3 buildAwsS3Client() {
+        return AmazonS3ClientBuilder
+                .standard()
+                .withRegion(getSystemProperty(
+                        "MEASURE_TRANSFER_S3_AWS_REGION"))
+                .withCredentials(new InstanceProfileCredentialsProvider(true))
+                .build();
+    }
+
+    public static PutObjectResult uploadMeasureDataToS3Bucket(MeasureTransferDTO ob, String measureId)
```
Comment (jkotanchik-SB) [inline src/main/java/mat/server/util/MeasureTransferUtil.java:23]: What's an `ob`?

Comment (adongare) [reply src/main/java/mat/server/util/MeasureTransferUtil.java]: initially, this was a generic method to upload any object but later I changed it to measuretransfer. didn't change the variable though

**Review by jkotanchik-SB — APPROVED** [review]

**Comment by adongare** [comment]
Thanks, Cecilia, Joe, Rohit for the review!

**Review by RohitKandimalla — APPROVED** [review]
